src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -348,9 +348,9 @@ public SSTableMultiWriter createSSTableMultiWriter(Descriptor descriptor,
                                                        Collection<Index.Group> indexGroups,
                                                        LifecycleNewTracker lifecycleNewTracker)
     {
-        ShardManager currentShardManager = getShardManager();
-        double flushDensity = realm.metrics().flushSizeOnDisk().get() * shardManager.shardSetCoverage() / currentShardManager.localSpaceCoverage();
-        ShardTracker boundaries = currentShardManager.boundaries(controller.getFlushShards(flushDensity));
+        maybeUpdateSelector();
```

Comment:
Nit: The local `currentShardManager` copy ensured that the instance won't change between the `flushDensity` calculation and the `boundaries` call. The updated version is racy.

(Yes, there was an error in the line below, it should have used `currentShardManager` for both.)

Reply:
Thanks for catching that. I see that we also use a direct reference to the `shardManager` in `UCS#getLevels`. I pushed a fix for that too.